addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusDataThingHandler.java

```diff
@@ -0,0 +1,798 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, String.format(
+                        "Seems to have writeTransformation but no other write parameters. Since the transformation did not return a JSON for command '%s' (channel %s), this is a configuration error",
+                        command, channelUID));
+                return;
+            } else {
+                transformedCommand = Transformation.tryConvertToCommand(transformOutput);
+                logger.trace("Converted transform output '{}' to command '{}' (type {})", transformOutput,
+                        transformedCommand.map(c -> c.toString()).orElse("<conversion failed>"),
+                        transformedCommand.map(c -> c.getClass().getName()).orElse("<conversion failed>"));
+            }
+        }
+
+        // We did not have JSON output from the transformation, so writeStart is absolute required. Abort if it is
+        // missing
+        if (writeStart == null) {
+            logger.warn(
+                    "Thing {} '{}': not processing command {} since writeStart is missing and transformation output is not a JSON",
+                    getThing().getUID(), getThing().getLabel(), command);
+            return;
+        }
+
+        if (!transformedCommand.isPresent()) {
+            // transformation failed, return
+            logger.warn("Cannot process command {} with channel {} since transformation was unsuccessful", command,
+                    channelUID);
+            return;
+        }
+
+        ModbusWriteRequestBlueprint request;
+        boolean writeMultiple = config.isWriteMultipleEvenWithSingleRegisterOrCoil();
+        if (config.getWriteType().equals(WRITE_TYPE_COIL)) {
+            Optional<Boolean> commandAsBoolean = ModbusBitUtilities.translateCommand2Boolean(transformedCommand.get());
+            if (!commandAsBoolean.isPresent()) {
+                logger.warn(
+                        "Cannot process command {} with channel {} since command is not OnOffType, OpenClosedType or Decimal trying to write to coil. Do not know how to convert to 0/1. Transformed command was '{}'",
+                        command, channelUID, transformedCommand);
+                return;
+            }
+            boolean data = commandAsBoolean.get();
+            request = new ModbusWriteCoilRequestBlueprintImpl(slaveId, writeStart, data, writeMultiple,
+                    config.getWriteMaxTries());
+        } else if (config.getWriteType().equals(WRITE_TYPE_HOLDING)) {
+            ModbusRegisterArray data = ModbusBitUtilities.commandToRegisters(transformedCommand.get(), writeValueType);
+            writeMultiple = writeMultiple || data.size() > 1;
+            request = new ModbusWriteRegisterRequestBlueprintImpl(slaveId, writeStart, data, writeMultiple,
+                    config.getWriteMaxTries());
+        } else {
+            // should not happen
+            throw new NotImplementedException();
```

Comment:
In Paper UI, it should be impossible to input anything else than a valid value.

With manually configured `things` file, inputting invalid values is possible. This is why the `writeType` parameter (and other parameters) is validated on initialization.

Since we have this validation in place, it should match `WRITE_TYPE_HOLDING`  (`"holding"`) or `WRITE_TYPE_COIL` (`"coil"`) always. 

In case new write types are added at some point, but this piece has been forgotten, the code would raise `NotImplementedException`, signaling incomplete implementation.

So it's more like a safe-guard for future.

Reply:
Wouldn't it make sense to validate the thing configuration once during the initialization of the thing (and set the thing to OFFLINE with reason CONFIGURATION_ERROR), so you can skip those checks at runtime?

Reply:
@triller-telekom as mentioned, this is actually checked during initialization and we do mark it as OFFLINE & CONFIGURATION_ERROR.

As said, the above check is just future-proofing the code, covering the all possibilities / state-space in the conditional. With future-proofing I mean that developer is "alerted" (clear crash) in case `writeType` is extended with additional values.

I have used elsewhere visitor pattern to guarantee similar future proofing (if new types / sub-classes are not handled they raise compliation error). Approach taken here makes things fail during runtime.

Please also note that checks cannot be skipped since the checks affect how the incoming command is processed. Or did I misunderstand?

Does this make sense?

Reply:
It somehow make sense to me but I agree with @martinvw that you should at least provide a message in the exception describing what went wrong.